keanu-project/src/main/java/io/improbable/keanu/util/csv/CsvWriter.java

```diff
@@ -0,0 +1,100 @@
+package io.improbable.keanu.util.csv;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
+
+public class CsvWriter {
+
+    private static final String DEFAULT_DELIMITER = ",";
+    private static final String DEFAULT_SUFFIX = ".csv";
+    private static final String NEWLINE = "\n";
+
+    private final List<List<String>> data;
+    private String delimiter;
+    private boolean writeHeader;
+    private List<String> header;
+    private FileWriter fileWriter;
+
+    public CsvWriter(List<List<String>> data) {
+        this.data = data;
+        this.delimiter = DEFAULT_DELIMITER;
+        this.writeHeader = false;
+        this.header = null;
+        this.fileWriter = null;
+    }
+
+    public File toFile(String filename, String directory) {
+        return toFile(filename, new File(directory));
+    }
+
+    public File toFile(String filename, File directory) {
+        try {
+            return toFile(File.createTempFile(filename, DEFAULT_SUFFIX, directory));
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Could not create: " + filename + " at: " + directory + ". " + e);
+        }
+    }
+
+    public File toFile(File file) {
+        try {
```

Comment:
To simplify things a bit we can use try-with-resources instead of old school try. Since the fileWriter is opened and closed within this method I don't see any reason to make it an instance variable. Also, a BufferedWriter is preferred I think. And if you create a toCsvLine(List<String> tokens) then we can really simply this method.

```
        try (BufferedWriter fileWriter = new BufferedWriter(new FileWriter(file))) {
            if (writeHeader) {
                writeHeader(fileWriter);
            }

            for (List<String> row : data) {
                fileWriter.append(toCsvLine(row));
                fileWriter.newLine();
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return file;
```

Reply:
So much simpler! changed